Add advanced kotlin blogpost (#98)
lunatech-labs/lunatech-blog: posts/2023-05-17-advanced-kotlin.adoc

= Embrace Kotlin: Where to Get Started for Scala Developers
AnastasiiaL
v1.0, 2023-03-31
:title: Exploring Advanced Kotlin Features: A Deep Dive for Scala Developers
:imagesdir: ../media/2023-03-31-embrace-kotlin
:lang: en
:tags: [scala, kotlin]

> Recently I shared my experience with trying out Kotlin coming from a Scala background. I talked about the syntax and some concept similarities.

Reviewer: Add a comma after _Recently_.

In this blog post, I want to focus on Kotlin's more advanced features and explore how Scala developers can leverage their existing knowledge to make the most out of Kotlin.

== Kotlin's Type System: A Closer Look for Scala Developers

When first diving into Kotlin from a Scala background, one of the most noticeable differences is the simplicity of Kotlin's type system.
While it's not as intricate as Scala's, it's robust and expressive, offering everything you need for most projects.

> In this section, we will delve deeper into Kotlin's type system, variance annotations, and type projections, and draw comparisons with Scala's type system.

Reviewer 2: Since you've already used "and" after type projections, I would suggest replacing the second "and" with "as well as"

Just like Scala, Kotlin is a statically-typed language. All types are known at compile-time, which means the compiler can catch type-related errors early.

> Unlike Scala, Kotlin does not have a universal superclass (like Any) and lower bounds, and it handles nullability differently.

Reviewer 3: `and` or `or` ?
```suggestion
Unlike Scala, Kotlin does not have a universal superclass (like Any) or lower bounds, and it handles nullability differently.

While Scala has _Option_ and _Either_ types to encapsulate optional values and potential failure cases, Kotlin tries to eliminate _NullPointerExceptions_ by distinguishing nullable and non-nullable types at the type system level.
Each type in Kotlin is either nullable (can hold a null value) or non-nullable (cannot hold a null value). This is indicated with a question mark suffix.
Here's a simple example:

Kotlin:
[source,kotlin]
----
var nonNullString: String = "Hello, World!"  // Non-Nullable String
var nullString: String? = null  // Nullable String
nonNullString = null // This line will cause a compile error
----

==== Variance Annotations
Variance annotations are Kotlin's solution to handling subtype relationships for generic types.
If you're familiar with Scala's + and - variance annotations, Kotlin's _out_ (covariant) and _in_ (contravariant) keywords will feel quite familiar.

However, Kotlin imposes stricter rules on variance to ensure type safety. In Kotlin, a type parameter marked as _out_ is only available in output positions (e.g., as the return type of a function) but not in input positions (e.g., as function parameters).

> Similarly, an _in_ parameter is only allowed in input positions. This safety measure helps prevent potential runtime _ClassCastException_ errors.

Reviewer 3: `at` could reduce the use of `in` - if reading out load one's tongue might trip a little, also there's a minor mix of singular and plural
```suggestion
Similarly, an _in_ parameter is only allowed at an input position. This safety measure helps prevent potential runtime _ClassCastException_ errors.
```

Or 

```suggestion
Similarly, _in_ parameters are only allowed at input positions. This safety measure helps prevent potential runtime _ClassCastException_ errors.

> Here's an example of using out (covariant) keyword in Kotlin:

Reviewer 3: ```suggestion
Here's an example of using the out (covariant) keyword in Kotlin:

[source,kotlin]
----
abstract class MyProducer<out T> {
    abstract fun produce(): T
}

fun demo(str: MyProducer<String>) {
    val objects: MyProducer<Any> = str // This is OK, since T is an out-parameter
    // ...
}
----

In this example, T is declared as an out parameter, so MyProducer<String> can be treated as MyProducer<Any>, respecting the subtyping relationship between String and Any.

> And an example for in (contravariant):

Reviewer 3: ```suggestion
And an example for the in (contravariant):

[source,kotlin]
----
abstract class MyConsumer<in T> {
    abstract fun consume(item: T)
}

fun demo(ints: MyConsumer<Int>) {
    val anyConsumer: MyConsumer<Any> = ints // Error: Type mismatch.
    val numberConsumer: MyConsumer<Number> = ints // This is OK, since T is an in-parameter
    // ...
}
----

> Here, T is declared as an in parameter, meaning MyConsumer<Int> can be treated as MyConsumer<Number>, again preserving the subtyping relationship.

Author: here I'm not sure "again preserving the subtyping relationship" can be a standalone sentence. It doesn't have a meaning without the first part🧐 unless it's okay in english and I'm just applying the rules from other languages

There are cases where a class works with its type parameter in both _in_ and _out_ positions.

This is where type projections come in. They let us define the variance for type parameters "on the fly", for specific instances of the class.
This allows us to work with such classes in a more flexible and safer way.
For example:

[source,kotlin]
----
class MyMutableList<T> {
    fun add(item: T) { /*...*/ }
    fun get(index: Int): T { /*...*/ }
}

fun printFirst(myList: MyMutableList<out Any>) {
    println(myList.get(0))
}

fun addToEnd(myList: MyMutableList<in String>, item: String) {
    myList.add(item)
}
----
In this example, MyMutableList works with T in both _in_ and _out_ positions, so we can't mark T as _in_ or _out_ in its declaration.
Instead, we can use type projections in the functions _printFirst_ and _addToEnd_.

Type projections in Kotlin serve a similar purpose as use-site variance in Java but are absent in Scala.

==== Comparing with Scala's Type System
While both Kotlin and Scala have powerful type systems, they each have unique features and handle some concepts differently.

> Scala's type system is more complex and powerful, with features such as higher-kinded types and implicit parameters, Scala can express complex abstractions that are not possible in Kotlin.

Reviewer 2: I would suggest a period after parameters and start the sentence anew with Scala
